Add methods to print a .list filfe with jira bugs

fehnkera opened the pull request:
Add method to print a .list file with jira bugs

Filters out bugs from the issue list and prints them in a separate bugs-jira.list file.
Uses the JIRA REST API to get more informations about an issue.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
add parser of jira bugs

Filters out the bugs from the issue list and prints them in a seperate
.list-file.
Additionaly to the normal issue attributes the 'resolution state', a
list of 'affected componentes' and a list of 'referenced issues' is
loaded from the xml-file and parsed in the .list-file.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

Commit:
adds jira api to load history

A method, that loads informations via the JIRA REST API for each issue,
is added. Especialy getting informations about the issue history is now
possible. Therefore the former inner method
merge_user_with_user_from_csv gets outsourced.
Due to increasing runtime caused by the API a flag 'with_api' is added
that makes it possible to disable the new functions which use the API.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -36,8 +36,13 @@
+
 reload(sys)
 sys.setdefaultencoding('utf-8')
+with_api = True  # flag to choose if information via api shall be loaded. Increases runtime
```
This flag needs to be added as part of the command-line API defined in Lines 49 to 52. How about renaming it to something like `retrieve_history`?

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -36,8 +36,13 @@
 reload(sys)
 sys.setdefaultencoding('utf-8')
+with_api = True  # flag to choose if information via api shall be loaded. Increases runtime
+project_url = 'https://issues.apache.org/jira'  # needed for api
```
Please retrieve this URL from the configuration (insert code around Line 60). The field is named `issueTrackerURL` (see [here](https://github.com/siemens/codeface/blob/7fb9c7a42e4ec3f7517acc41f01601106b2baae6/conf/spark.conf#L24), for example). Then, you need to pass it to the respective functions if necessary.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -168,6 +187,19 @@ def merge_user_with_user_from_csv(user, persons):
             issue['projectId'] = project.attributes['id'].value
 
+            resolution = issue_x.getElementsByTagName('resolution')[0]
+            issue['resolution'] = resolution.firstChild.data
```
Is there something similar available for the field `state`?

fehnkera replied:
What exactly do you mean with something similar available? The informations about the state where already loaded before my changes. This happens analogous.

clhunsen replied:
Ah, I must have overlooked the retrieval of the `state` attribute (Line 184). :man_facepalming: So, nevermind, everything is fine here.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -225,6 +259,69 @@ def merge_user_with_user_from_csv(user, persons):
+        changelog = api_issue.changelog
+        histories = list()
+
+        for change in changelog.histories:
```
Please put a comment here about the time order in which the history entries are visited here: newest to oldest, oldest to newest, random? Without this information, it is hard to understand the logic in the following code parts.

fehnkera replied:
history entries are checked in time order from oldest to newest. I'll put a comment about this in the code

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -225,6 +259,69 @@ def merge_user_with_user_from_csv(user, persons):
+                date = change.created
+                d = datetime.strptime(date, "%Y-%m-%dT%H:%M:%S.%f+0000")
+                history['date'] = d.strftime('%Y-%m-%d %H:%M:%S')
+                histories.append(history)
```
I suggest to add `history` to `histories` at the end of this for-loop, just for consistency and comprehensibility.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -225,6 +259,69 @@ def merge_user_with_user_from_csv(user, persons):
+                histories.append(history)
+
+                # updates state and resolution for issue and comments in the past
+                if (min_date <= issue['creationDate']) & (issue['creationDate'] < history['date']):
```
I am not sure what you want to achieve in this if-block and whether the condition is appropriate for this task. Can you please elaborate?

fehnkera replied:
The if-block examines if the issue creation date is between the last checked history entry and the now checked history entry (history entries are checked in time order from oldest to newest). If this is true the state and resolution of the issue get updated, so that you get the state and the resolution the issue had when it was created. I'm not sure if this is necessary for the issue itself, because there might be default values after creating an issue.

clhunsen replied:
As discussed in our Thursday meeting, I would like to track this information differently: the global information about the issue as one part and the event information as the other part. The current state of the issue needs to be incorporated in the latter. (see also the comments on Line 498 below)

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -225,6 +259,69 @@ def merge_user_with_user_from_csv(user, persons):
+
+                    if (min_date <= comment['changeDate']) & (comment['changeDate'] < history['date']):
+                        comment['state'] = old_state
+                        comment['resolution'] = old_resolution
```
Can you please elaborate how a comment can have a resolution and a state? Are these properties independent of the issue's properties? Do the reflect each other because you can change the state/resolution of an issue with the help of a comment?

fehnkera replied:
The state and the resolution of an comment shall display what state and resolution the issue had when the comment was written. This is elaborated with the help of the history.
In my opinion this is an interesting information about a comment. But I recognize that this attributes could have better names.

clhunsen replied:
With your comments and the information from our discussion, this makes sense to me now. A renaming is not necessary, but may help later.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -225,6 +259,69 @@ def merge_user_with_user_from_csv(user, persons):
+
+                for comment in issue['comments']:
+
+                    if (min_date <= comment['changeDate']) & (comment['changeDate'] < history['date']):
```
Same holds here: Not sure why we need to have this if-block here and what you want to achieve with it.

fehnkera replied:
Analogous to the if-block above. This Block elaborates what state and resolution the issue had when the comment was written. In my opinion this is an interesting information, because it makes a difference if you comment on an resolved bug or on an unresolved bug. If we don't want this information this if-block is unnecessary

clhunsen replied:
Thanks for giving me a rationale. :+1: The information is strongly needed, but needs to be incorporated solely in the event part of the data (see also the comments on Lines 310 and 498).

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -330,6 +427,91 @@ def print_to_disk(issues, results_folder):
+                for history in issue['history']:
+                    lines.append((
+                        issue['externalId'],
+                        history['state'],
```
When consolidating the JIRA and GitHub issues, we need to think carefully about statements like this. For example, I would like to stick with the issue's latest state here, thus, we would need to put the change's state in a later column. (Same holds for the comment's state in Line 481 above.)

How about adding an additional column so that we can reflect old and new value in two distinct columns.

fehnkera replied:
Good point.
I think it is an good idea to have the global issue state/ resolution in columns at the beginning and the information about the state/ resolution on creation date in  different columns at the the end. What do you think about this?

clhunsen replied:
Yes, that is a really good idea. Try to find (in a later step) a format that supports various information for different types of events (see also the GitHub stuff).

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -330,6 +427,91 @@ def print_to_disk(issues, results_folder):
+                issue['references'],
+                "open",  ## event.name
+                issue['resolution'],
+                issue['components']
```
How does the CSV writer stores the components here? Comma-separated, as embedded JSON, ...?

fehnkera replied:
The components (as well as the references) get written in a list separated by a comma. Furthermore they get written as an Unicode-String.
This is agreed with Barbara but if you have an better idea I can easily change it.

clhunsen replied:
Okay, I just wanted to check whether Python behaves correctly here so that we can parse the data item later.

clhunsen reviewed (commented):
Hi @fehnkera,

thank you very much for your changes. I annotated three types of things with my comments:
- spots in the code that I do not understand, but that can be improved (easily) with more inline comments,
- spots in the code that need to be improved (e.g., the CLI), and
- spots in the code that we need to keep in mind when consolidating the JIRA issues with the GitHub stuff, or that we may think about some more time whether we really want to implement it that way.

Feel free to comment or talk to me personally. Do not hesitate to ask if you do not understand some of my comments.

bockthom commented:
@fehnkera In addition to the comments from Claus, please don't forget to update the Copyright header in the file you have changed (i.e., insert a new line containing your name and e-mail address in the Copyright header...)

Commit:
apply review on previous commits

The review from clhunsen and bockthom on the previous two commits gets
applied. Especially following things change: The 'with_api' flag is
renamed to 'skip_history' and is now part of the  command-line API. The
issue-url is now loaded from the configuration file. The information
order in the output-file changed. Now there are fields for the global
state and resolution and fields for the state and resolution on
creation. Issue state and resolution on creation now has default values.
The API informations get adjust to the XML-informations. The copyrigth
header gets updated.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -36,15 +37,19 @@
     parser.add_argument('-c', '--config', help="Codeface configuration file", default='codeface.conf')
     parser.add_argument('-p', '--project', help="Project configuration file", required=True)
     parser.add_argument('resdir', help="Directory to store analysis results in")
+    parser.add_argument('-s', '--skip_history', help='Skips methods that retrieve the history', action='store_true')
```
I would like to add some more context information in the help text here what "retrieve" means exactly. I propose something like this: "Skip methods that retrieve additional history information from the configured JIRA server".

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -330,6 +429,99 @@ def print_to_disk(issues, results_folder):
+    """Sorts of bug issues and prints them to file 'bugs-jira.list' in result folder
+
+    :param issues: the issues to sort of bugs
+    :param results_folder: the folder where to place 'bugs-jira.list' output file
```
The parameter `skip_history` is missing from the documentation.

clhunsen commented on issue_processing/jira_issue_processing.py:
```diff
@@ -67,15 +72,20 @@ def run():
-    # 3) update user data with Codeface database
+    # 3) load issue information via api
+    if not args.skip_history:
+        load_issue_via_api(issues, persons, __conf)
```
Please pass the actual issue-tracker URL here, not the entire configuration object `__conf` as it is not needed here actually.

clhunsen reviewed (approved):
Apart from some minor things, this PR looks fine for me now. Nice work, @fehnkera. :+1: 

Please work on the last comments, then I will merge. :grinning:

Commit:
Minor changes after second review

After another review from @clhunsen following things have changed: The
help-documentation of 'skip_history'got more detailed. The method
'load_issue_via_api' only gets the project-url instead of the complete
config. The parameter documentation for 'skip_history' in
'print_to_disk_bugs' is added as it was missing before.

Signed-off-by: Anselm Fehnker <fehnker@fim.uni-passau.de>
